### PR TITLE
Prepared-metadata reuse for collective-free async checkpoint saves

### PR DESCRIPTION
**Pull request by asolergi-nv**
# Reuse a prepared `.metadata` across checkpoint saves (collective-free save)

This is a **linked pair of PRs** — both must land for the feature to work:

- **NVRx** (`asolergi-nv/nvidia-resiliency-ext`, branch `feat/ckpt-metadata-reuse`): adds the backend reuse path to the async saver.
- **Megatron-LM** (`asolergi-nv/Megatron-LM`, branch `feat/ckpt-metadata-reuse`): adds the `--ckpt-metadata` / `--ckpt-metadata-create` knobs that drive it.

The Megatron PR depends on the NVRx PR (it passes the new `reuse_metadata_obj` / `create_metadata_path` arguments). Merge/release NVRx first.

---

## Problem

On the NVRx async save path, every distributed checkpoint save performs two
cross-rank collectives that scale with world size and are pure coordination
(no model data):

1. **Planning** — `gather_object(local_plan)` to the coordinator to build the
   global `Metadata`.
2. **Finalize** — `gather_object(write_results)` to assemble and write `.metadata`.

For large jobs these gathers are a non-trivial, recurring tax on every save,
even though the checkpoint **structure is constant** across an entire run
(same model, parallelism, and world size). The global `Metadata` is therefore
identical save-to-save and across resumes — it is wasteful to rebuild it every
time via collectives.

## Solution

Allow a **prepared, complete `Metadata`** object to be supplied to the save, so
both gathers are skipped entirely:

### NVRx side (`async_ckpt/state_dict_saver.py`)
- `save_state_dict_async_plan(..., reuse_metadata_obj=None)`: when a prepared
  metadata is supplied, the decentralized planning branch skips **both** the
  `verify_global_md_reuse` all-reduce and the `gather_object(local_plan)` —
  planning becomes purely local.
- `save_state_dict_async_finalize(..., reuse_metadata_obj=None, create_metadata_path=None)`:
  the reuse path skips `gather_object(write_results)` and instead detects write
  failures with a **single `all_reduce(MIN)` of a 1-element flag**, then writes
  the prepared metadata verbatim as `.metadata`.
- **Create mode** (`create_metadata_path`): the first save runs the normal
  (collective) path, then the coordinator reads back the just-written complete
  metadata, broadcasts it, and persists it to a container-local file so later
  saves in the job/run can reuse it. Returns the complete metadata.

A single, tiny `all_reduce(MIN)` is intentionally retained on the reuse path for
write-failure safety (so a failed rank still aborts the whole job).

### Megatron side
- `--ckpt-metadata <FILE>`: reuse mode — load the prepared `.metadata` (pickle)
  and hand it to the save strategy as `prepared_metadata` (skips both gathers).
- `--ckpt-metadata-create`: first-job mode — the first save creates the
  `<FILE>` via the collective path, then subsequent saves reuse it.
- `TorchDistSaveShardedStrategy` gains `prepared_metadata` / `ckpt_metadata_create_path`
  and passes `reuse_metadata_obj` / `create_metadata_path` through to NVRx; the
  async `finalize_fn` captures the returned complete metadata in create mode.

## Save-side collectives in NVRx (and which disappear)

Each NVRx async save (decentralized plan path) triggers these cross-rank
collectives — none of which move model/optimizer *data*; they only coordinate
the checkpoint's `Metadata`:

| Phase | Collective | Purpose |
|---|---|---|
| Planning (`save_state_dict_async_plan`) | `all_reduce` inside `verify_global_md_reuse` | check whether the previously-loaded plans still match (only when prior plans were loaded) |
| Planning | `gather_object(local_plan)` → coordinator | gather every rank's local plan to build the global `Metadata` |
| Finalize (`save_state_dict_async_finalize`) | `gather_object(write_results)` → coordinator | gather every rank's write results to assemble and write `.metadata` |
| Finalize | `broadcast(failures_occurred)` (1 int) | propagate write-failure status to all ranks |

### Which disappear after the first save(s), and why

- **Planning `all_reduce` + `gather_object(local_plan)`** — eliminated by NVRx's
  own `CheckpointMetadataCache`. With `--ckpt-assume-constant-structure`, the cache
  caches the central plan and, once it has observed an *identical* structure on
  **two** consecutive saves, sets `validated_cache_reuse=True` and reuses the
  cached central plan. So these run on save **#1 and #2** (the cache needs two
  matching observations to validate) and then vanish from save **#3** onward.
- **Finalize `gather_object(write_results)` + `broadcast`** — *not* cached by
  stock NVRx: the per-rank write results are re-gathered to rebuild `.metadata`
  on **every** save, so this collective never disappears on its own.

### What `--ckpt-metadata` changes

Supplying a prepared, complete `Metadata` lets the save skip **both** gathers:

- **Reuse mode**: `reuse_metadata_obj` is set → the planning branch skips the
  `verify_global_md_reuse` `all_reduce` *and* `gather_object(local_plan)`
  (planning is purely local), and the finalize skips `gather_object(write_results)`,
  replacing it with a single `all_reduce(MIN)` of a 1-element flag for
  write-failure detection. This holds from the **very first** reuse save — no
  warm-up saves needed — and is independent of `CheckpointMetadataCache`.
- **Create mode** (first job): runs the full collective path once to produce the
  metadata, plus a one-time `broadcast` of the complete metadata to all ranks,
  then persists it; every later save reuses it collective-free.

Net steady-state: the only remaining save-side collective is the 1-element
`all_reduce(MIN)` failure check (kept by design). (The separate mcore-level
collectives — the FullyParallel save-distribution `all_gather_object` and the
sharding-integrity `determine_global_metadata` gather — are addressed by the
sibling Megatron PRs M4 and M5 and are out of scope here.)

## Correctness constraints

Reuse is only valid when the run that **reuses** a `.metadata` has the **same
checkpoint structure, world size, and `--dist-ckpt-workers`** as the run that
**created** it (the caller owns this guarantee). In particular the prepared
metadata must come from a save done with the **same `--keep-redundant-extra-state`
setting**, since that toggle changes the saved item set and per-file byte
offsets. A stale/mismatched prepared metadata produces a wrong shard map and a
load-time unpickling error.

## Usage

First job (create once):
```
--async-save --use-persistent-ckpt-worker \
--ckpt-metadata $CONTAINER/prepared.metadata --ckpt-metadata-create
```
Subsequent jobs (reuse, collective-free saves):
```
--async-save --use-persistent-ckpt-worker \
--ckpt-metadata $CONTAINER/prepared.metadata
```

## Result

With `--ckpt-metadata` reuse active (and constant structure), the only remaining
save-side collective is the 1-element `all_reduce(MIN)` failure check. Both
`gather_object`s are eliminated.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
🔗 **Paired PR** (must land together): Megatron `--ckpt-metadata` knobs that drive this backend → https://github.com/NVIDIA/Megatron-LM/pull/5555

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py [hunk 1]
```diff
@@ -427,11 +514,40 @@ def save_state_dict_async_finalize(
...
+        )
+        torch.distributed.all_reduce(
+            flag, op=torch.distributed.ReduceOp.MIN, group=dist_wrapper.group
+        )
+        if flag.item() != 1:
+            node_failures = {} if local_ok else {dist_wrapper.get_rank(): write_results}
+            raise CheckpointException("write", node_failures)
+        if dist_wrapper.is_coordinator:
+            write_start = time()
+            _write_metadata_only(storage_writer, reuse_metadata_obj)
```
Comment (shurkat-nvidia) [inline src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:715]: How are you ensuring that the length/offset are correctly collected from WriteResult reported by every rank here?

File: tests/checkpointing/unit/test_async_writer.py [hunk 2]
```diff
@@ -173,6 +173,131 @@ def test_async_is_equivalent_to_sync(
...
+            )
+
+        async_request = AsyncRequest(
+            save_fn, save_args, [finalize_fn], preload_fn=preload_fn, async_fn_kwargs={}
+        )
+        async_queue.schedule_async_request(async_request)
+        async_queue.maybe_finalize_async_calls(blocking=True, no_dist=False)
+        return finalize_result.get('metadata')
+
+    def test_prepared_metadata_reuse_roundtrip(self, tmp_path_dist_ckpt):
```
Comment (shurkat-nvidia) [inline tests/checkpointing/unit/test_async_writer.py:232]: You should replace model weights by random weights between the metadata reuse. That will check that the metadata reuse holds for multiple training iterations.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py (hunk 1 above)
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:715]: Good question — this is the core invariant the reuse path relies on, so let me be explicit. The per-chunk `offset`/`length` in `storage_data` are a function of the checkpoint **structure** — item shapes/dtypes and the write-bucketing (which depends on world size and `--dist-ckpt-workers`) — not of the tensor **values**. For a fixed structure the byte layout is therefore identical save-to-save, which is exactly what makes a prepared metadata reusable.

I've made the unit test demonstrate this directly (also addressing your other comment): it now re-randomizes every parameter between the `create` save and the `reuse` save, then asserts the reuse checkpoint — whose `.metadata` is the verbatim prepared metadata built from the *old* weights — loads back to the *new* weights. If any offset/length depended on the data, that round-trip would fail. The validity constraint (same structure / world size / `--dist-ckpt-workers` / `--keep-redundant-extra-state`) is documented in the docstring and PR body and is the caller's responsibility.

If you'd like a stronger guarantee we could add a purely-local save-time check that compares this rank's actual `WriteResult` offsets/lengths against the prepared metadata (no collective needed) and fails fast on a mismatch — happy to add that as a follow-up if you think it's worth it. Addressed in 7ce102906.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py [hunk 3]
```diff
@@ -414,9 +425,85 @@ def verify_global_md_reuse(
...
+    with fs.create_stream(tmp_path, "wb") as f:
+        pickle.dump(metadata, f)
+        try:
+            os.fsync(f.fileno())
+        except (AttributeError, OSError):
+            os.sync()
+    metadata_path = storage_writer._get_metadata_path()
+    if fs.exists(metadata_path):
+        fs.rm_file(metadata_path)
+    fs.rename(tmp_path, metadata_path)
```
[the remark this answers is not in the document]
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:478]: Agreed — fixed in 7ce102906. Added `f.flush()` before `os.fsync(f.fileno())` so the pickle tail sitting in the file object's userspace buffer is drained to the OS before the fsync, matching the sibling `_atomic_pickle_write`.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py [hunk 4]
```diff
@@ -414,9 +425,85 @@ def verify_global_md_reuse(
...
+    import os
+    import pickle
+
+    if getattr(storage_writer, "use_msc", False):
+        import multistorageclient as msc
+
+        path = os.path.join(storage_writer.checkpoint_dir, ".metadata")
+        with msc.open(path, "wb") as f:
+            pickle.dump(metadata, f)
+        return
```
[the remark this answers is not in the document]
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:459]: I'm intentionally leaving the MSC branch as-is, for consistency with the stock writer: `FileSystemWriterAsync.finish()` also writes `.metadata` directly via `msc.open(...)` with no temp+rename (`filesystem_async.py:1237-1257`), and this helper mirrors that path. MSC/object-store writers finalize the object on `close()` (multipart-complete), so an aborted write generally doesn't publish a half-object the way a partial POSIX file write would. Making only the reuse path atomic would diverge from `finish()` and depend on per-backend rename semantics; if we want atomic MSC metadata it should be a separate change applied to both `finish()` and `_write_metadata_only` so they stay aligned.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py [hunk 5]
```diff
@@ -427,11 +514,40 @@ def save_state_dict_async_finalize(
...
+        local_ok = 1 if isinstance(write_results, list) else 0
+        flag = torch.tensor(
+            [local_ok], dtype=torch.int, device=torch.cuda.current_device()
+        )
+        torch.distributed.all_reduce(
+            flag, op=torch.distributed.ReduceOp.MIN, group=dist_wrapper.group
+        )
+        if flag.item() != 1:
+            node_failures = {} if local_ok else {dist_wrapper.get_rank(): write_results}
+            raise CheckpointException("write", node_failures)
```
[the remark this answers is not in the document]
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py]: Good catch — addressed in 7ce102906. On the reuse path a rank whose own write succeeded but that sees the `all_reduce(MIN)` flag trip now logs a warning pointing operators to the failing rank's logs before raising (the failing rank still raises with its own traceback). I also documented this reduced-detail behavior in the docstring. Collecting full per-rank detail on the coordinator would require exactly the gather this path is designed to skip, so the warning plus the failing rank's own exception is the intended trade-off.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py [hunk 6]
```diff
@@ -427,11 +514,40 @@ def save_state_dict_async_finalize(
...
+            coordinator, which this metadata already has) and writes this object
+            verbatim as the checkpoint's ``.metadata``. Write failures are still
+            surfaced, but via a single ``all_reduce`` of a success flag instead
+            of gathering the (large) per-rank write-result lists. Valid only when
+            the structure / world size / dist-ckpt-workers match the run that
+            produced the prepared metadata (the caller owns this guarantee).
 
     Returns: None
     """
     write_results = storage_writer.retrieve_write_results()
```
[the remark this answers is not in the document]
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py]: Fixed in 7ce102906 — updated the `Returns` section to `Optional[Metadata]` (complete metadata on all ranks in create mode, otherwise `None`) and documented the `create_metadata_path` argument.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (asolergi-nv) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:478]: Agreed — fixed in 7ce102906. Dropped the `if fs.exists(...): fs.rm_file(...)` and rename directly; POSIX `rename(2)` atomically replaces an existing destination, so this removes the window where `.metadata` was absent and makes the write match the docstring's “atomic” claim.

File: tests/checkpointing/unit/test_async_writer.py (hunk 2 above)
Comment (asolergi-nv) [reply tests/checkpointing/unit/test_async_writer.py:232]: Done in 7ce102906. The test now re-randomizes every parameter between the `create` and `reuse` saves (emulating a later training iteration — same structure, new data) and asserts the reuse checkpoint loads back to the *new* weights. This exercises reuse across changed data and, as noted on your other comment, demonstrates that the reused offsets/lengths depend only on structure, not on the values.

File: src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py (hunk 1 above)
Comment (shurkat-nvidia) [reply src/nvidia_resiliency_ext/checkpointing/async_ckpt/state_dict_saver.py:715]: One potential idea, instead of a fail-fast would be to hijack the mechanism that records whether ranks are done writing. Currently, that's an all-reduce/all-gather of a single value tensor, which can be 0 or 1. We could add a value 2 that indicates the rank was done writing, but its offsets/lengths/something else about distcp file has changed and the metadata cache can't be reused.

**Comment by shurkat-nvidia** [comment]
My main concern with this PR is that there's already a path to provide cached metadata to NVRX, via `CheckpointMetadataCache`. If we can address the integration issues in Megatron-LM, we won't need this additional code in NVRX.
